python_modules/dagster/dagster/_core/definitions/assets.py

```diff
@@ -1060,14 +1060,20 @@ def infer_partition_mapping(
                 partition_mapping, self._partitions_def, upstream_partitions_def
             )
 
+    def has_output_for_asset_key(self, key: AssetKey) -> bool:
+        return key in self.output_name_by_asset_key
+
     def get_output_name_for_asset_key(self, key: AssetKey) -> str:
-        for output_name, asset_key in self.keys_by_output_name.items():
-            if key == asset_key:
-                return output_name
+        try:
+            return self.output_name_by_asset_key[key]
+        except KeyError:
+            raise DagsterInvariantViolationError(
+                f"Asset key {key.to_user_string()} not found in AssetsDefinition"
+            )
 
-        raise DagsterInvariantViolationError(
-            f"Asset key {key.to_user_string()} not found in AssetsDefinition"
-        )
+    @cached_property
+    def output_name_by_asset_key(self) -> Mapping[AssetKey, str]:
+        return {key: output_name for output_name, key in self.keys_by_output_name.items()}
```

Comment:
cache this to avoid pain on large multi-assets

Reply:
After [this PR](https://github.com/dagster-io/dagster/pull/22554), you can use `self._computation.output_names_by_key`

Reply:
That PR has been merged, so I think this comment is still relevant and we should use `output_names_by_key`.